sw/device/lib/msg_print.c

```diff
@@ -0,0 +1,107 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/msg_print.h"
+
+#include <stdarg.h>
+
+static void print_digit(print_char_func print_char, unsigned int d) {
+  print_char("0123456789ABCDEF"[d]);
+}
+
+static void print_num(print_char_func print_char, int w, unsigned int n,
+                      unsigned int base) {
+  if (--w > 0 || n >= base) {
+    print_num(print_char, w, n / base, base);
+  }
+  print_digit(print_char, n % base);
+}
+
+void msg_print(print_char_func print_char, const char *fmt, ...) {
+  char ch;
+  va_list va;
+
+  va_start(va, fmt);
+  while ((ch = *fmt) != '\0') {
+    ++fmt;
+    if (ch != kFormatSpecifier) {
+      // Add CR to new line automatically.
+      if (ch == '\n') {
+        print_char('\r');
+      }
+      print_char(ch);
+    } else {
+      int w = 0;
+      while ((ch = *fmt) != '\0') {
+        ++fmt;
+        // Parse width field.
+        if (ch >= '0' && ch <= '9') {
+          w = w * 10 + (ch - '0');
+          continue;
+        } else {
+          switch (ch) {
+            case '\0': {
+              break;
+            }
+            case kBinary: {
+              unsigned int n = va_arg(va, unsigned int);
+              print_num(print_char, w, n, 2);
+              break;
+            }
+            case kDecimalI:
+            case kDecimal: {
+              unsigned int n = va_arg(va, unsigned int);
+              if ((int)n < 0) {
+                print_char('-');
+                n = -n;
+              }
+              print_num(print_char, w, n, 10);
+              break;
+            }
+            case kUnsigned: {
+              unsigned int n = va_arg(va, unsigned int);
+              print_num(print_char, w, n, 10);
+              break;
+            }
+            case kOctal: {
+              unsigned int n = va_arg(va, unsigned int);
+              print_num(print_char, w, n, 8);
+              break;
+            }
+            case kHexLower:
+            case kHexAltLower: {
+              // TODO: This will still print in upper case.
+              unsigned int n = va_arg(va, unsigned int);
+              print_num(print_char, w, n, 16);
+              break;
+            }
+            case kHexUpper:
+            case kHexAltUpper: {
+              unsigned int n = va_arg(va, unsigned int);
+              print_num(print_char, w, n, 16);
+              break;
+            }
```

Comment:
These are all doing the same thing; re-use the same code block for them, but keep the TODO.

Reply:
Done.